cmd/kubeadm/app/util/users/users_linux.go

```diff
@@ -0,0 +1,631 @@
+			if err != nil {
+				return nil, errors.Wrapf(err, "could not parse value for %s at line %d", k, i)
+			}
+			*v = val
+		}
+	}
+	return &l, nil
+}
+
+// parseEntries can be used to parse an /etc/passwd or /etc/group file as their format is similar.
+// It returns a slice of entries obtained from the file.
+// https://www.cyberciti.biz/faq/understanding-etcpasswd-file-format/
+// https://www.cyberciti.biz/faq/understanding-etcgroup-file/
+func parseEntries(file string, totalFields int) ([]*entry, error) {
+	if totalFields != totalFieldsUser && totalFields != totalFieldsGroup {
+		return nil, errors.Errorf("unsupported total fields for entry parsing: %d", totalFields)
+	}
+	lines := strings.Split(file, "\n")
+	entries := []*entry{}
+	for i, line := range lines {
+		line = strings.TrimSpace(line)
+		if len(line) == 0 {
+			continue
+		}
+		fields := strings.Split(line, ":")
+		if len(fields) != totalFields {
+			return nil, errors.Errorf("entry must have %d fields separated by ':', "+
+				"got %d at line %d: %s", totalFields, len(fields), i, line)
+		}
+		id, err := strconv.ParseInt(fields[2], 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error parsing id at line %d", i)
+		}
+		entry := &entry{name: fields[0], id: id}
+		if totalFields == totalFieldsGroup {
+			entry.userNames = strings.Split(fields[3], ",")
+		} else {
+			gid, err := strconv.ParseInt(fields[3], 10, 64)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error parsing GID at line %d", i)
+			}
+			entry.gid = gid
+			entry.shell = fields[6]
+		}
+		entries = append(entries, entry)
+	}
+	return entries, nil
+}
+
+// validateEntries takes user and group entries and validates if these entries are valid based on limits,
+// mapping between users and groups and specs. Returns slices of missing user and group entries that must be created.
+// Returns an error if existing users and groups do not match requirements.
+func validateEntries(users, groups []*entry, limits *limits) ([]*entry, []*entry, error) {
+	u := []*entry{}
+	g := []*entry{}
+	// Validate users
+	for _, uc := range usersToCreateSpec {
+		for _, user := range users {
+			if uc.name != user.name {
+				continue
+			}
+			// Found existing user
+			if user.id < limits.minUID || user.id > limits.maxUID {
+				return nil, nil, errors.Errorf("UID %d for user %q is outside the system UID range: %d - %d",
+					user.id, user.name, limits.minUID, limits.maxUID)
+			}
+			if user.shell != noshell {
+				return nil, nil, errors.Errorf("user %q has unexpected shell %q; expected %q",
+					user.name, user.shell, noshell)
+			}
+			for _, g := range groups {
+				if g.id != user.gid {
+					continue
+				}
+				// Found matching group GID for user GID
+				if g.name != uc.name {
+					return nil, nil, errors.Errorf("user %q has GID %d but the group with that GID is not named %q",
+						uc.name, g.id, uc.name)
+				}
+				goto skipUser // Valid group GID and name; skip
```

Comment:
skipping a loop cycle in the parent loop using a goto is not so uncommon:
https://kodify.net/go/continue-loop/#example-continue-with-a-labelled-for-loop

it's less LoC too.

Reply:
I only see one instance where it is used in non test code in k/k https://github.com/kubernetes/kubernetes/search?q=goto

It seems like kubeadm already uses this pattern so its ok to do so here.

Reply:
i know...
i actually advocate for gotos, since that's something that Linux kernel code review recommend too.


they discourage too many nested indentations:
```
for ... {
	if ... {
		if ... {
...
```

and instead tell contributors to use a goto.